### PR TITLE
half join WIP

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: hydroflow/src/compiled/pull/symmetric_hash_join.rs [hunk 1]
```diff
@@ -44,39 +42,39 @@ where
...
                 // TODO(justin): unnecessary clone (sometimes).
                 let result = (k.clone(), (vs.pop().unwrap(), v.clone()));
                 if vs.is_empty() {
-                    self.state.rbuffer = None;
+                    self.state.1.buffer = None;
                 }
                 return Some(result);
             }
 
             if let Some((k, v1)) = self.lhs.next() {
```
Comment (jhellerstein) [inline hydroflow/src/compiled/pull/symmetric_hash_join.rs:61]: could be DRYer -- both left and right use the same code, just different arguments. Factor out the logic?

File: hydroflow_lang/src/graph/ops/halfjoin.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    parse_single_lifetime, OperatorConstraints, OperatorWriteOutput, Persistence, WriteContextArgs,
+    WriteIteratorArgs, RANGE_1,
+};
+
+use quote::quote_spanned;
+use syn::parse_quote;
+
+/// > 2 input streams of type <(K, V1)> and <(K, V2)>, 1 output stream of type <(K, (V1, V2))>
+///
+/// Forms the "streaming half-join" of the tuples in the input streams by their first (key)
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/halfjoin.rs:11]: This should not be an op in the surface syntax, should it? 

Either way, the documentation here is not on target.

Comment (jhellerstein) [reply hydroflow_lang/src/graph/ops/halfjoin.rs:11]: If this is the canonical relational Join operator, don't call the file `halfjoin.rs`?

File: hydroflow_lang/src/graph/ops/halfjoin.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                 wi @ &WriteIteratorArgs { ident, inputs, .. },
+                 diagnostics| {
+        let persistence = parse_single_lifetime(wi, diagnostics).unwrap_or(Persistence::Static);
+
+        let joindata_ident = wc.make_ident("joindata");
+        let borrow_ident = wc.make_ident("joindata_borrow");
+        let (init, borrow) = match persistence {
+            Persistence::Epoch => (
+                quote_spanned! {op_span=>
+                    #root::lang::monotonic_map::MonotonicMap::new_init(
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/halfjoin.rs:86]: interesting how the shorter-lived variant is the wrapped-up structure!

File: hydroflow_lang/src/graph/ops/join.rs [hunk 4]
```diff
@@ -16,6 +17,46 @@ use syn::parse_quote;
 /// source_iter(vec![("hello", "cleveland")]) -> [1]my_join;
 /// my_join -> for_each(|(k, (v1, v2))| println!("({}, ({}, {}))", k, v1, v2));
 /// ```
+///
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/join.rs:20]: should the join docs from `halfjoin.rs` go here?

**Review by jhellerstein — COMMENTED** [review]
some notes on code comments and structure inline

**Comment by MingweiSamuel** [comment]
#281

**Pull request closed without merging**